test/runtests.jl

```diff
@@ -1,3 +1,135 @@
+include("JuMP_sparse_linear_mpc.jl")
 
-@test true
+
+nt = 3 # number of time steps
+ns = 2 # number of states
+nu = 1 # number of inputs
+
+# generate random Q, R, A, and B matrices
+Random.seed!(10)
+Q_rand = Random.rand(ns,ns)
+Q = Q_rand * transpose(Q_rand) + I
+R_rand   = Random.rand(nu,nu)
+R    = R_rand * transpose(R_rand) + I
+
+A_rand = rand(ns, ns)
+A = A_rand * transpose(A_rand) + I
+B = rand(ns, nu)
+
+# generate upper and lower bounds
+lvar = rand(0:.0000001:1, (ns*nt + nu*nt))
+uvar = lvar .+ 10
+
+
+# build JuMP models
+mnb  = build_QP_JuMP_model(Q,R,A,B, nt)
+mlb  = build_QP_JuMP_model(Q,R,A,B, nt; lvar = lvar)
+mub  = build_QP_JuMP_model(Q,R,A,B, nt; uvar= uvar)
+mulb = build_QP_JuMP_model(Q,R,A,B, nt; lvar = lvar, uvar= uvar)
+
+# Build Quadratic Model
+qpnb  = get_QM(Q, R, A, B, nt)
+qplb  = get_QM(Q, R, A, B, nt; lvar=lvar)
+qpub  = get_QM(Q, R, A, B, nt; uvar=uvar)
+qpulb = get_QM(Q, R, A, B, nt; lvar = lvar, uvar=uvar)
+
+
+# Solve JuMP model with Ipopt
+optimize!(mnb)
+optimize!(mlb)
+optimize!(mub)
+optimize!(mulb)
+
+# Solve QP with Ipopt
+ipopt_nb  = ipopt(qpnb)
+ipopt_lb  = ipopt(qplb)
+ipopt_ub  = ipopt(qpub)
+ipopt_ulb = ipopt(qpulb)
+
+# Solve QP with MadNLP
+madnlp_nb  = madnlp(qpnb, max_iter=50)
+madnlp_lb  = madnlp(qplb, max_iter=50)
+madnlp_ub  = madnlp(qpub, max_iter = 50)
+madnlp_ulb = madnlp(qpulb, max_iter = 50)
+
+
+# Test results
+@test abs(objective_value(mnb) - ipopt_nb.objective)   < 1e-7
+@test abs(objective_value(mlb) - ipopt_lb.objective)   < 1e-7
+@test abs(objective_value(mub) - ipopt_ub.objective)   < 1e-7
+@test abs(objective_value(mulb) - ipopt_ulb.objective) < 1e-7
+
+@test abs(objective_value(mnb) - madnlp_nb.objective)   < 1e-7
+@test abs(objective_value(mlb) - madnlp_lb.objective)   < 1e-7
+@test abs(objective_value(mub) - madnlp_ub.objective)   < 1e-7
+@test abs(objective_value(mulb) - madnlp_ulb.objective) < 1e-7
+
+
+# Add Qf matrix
+Qf_rand = Random.rand(ns,ns)
+Qf = Qf_rand * transpose(Qf_rand) + I
+
+mulbf  = build_QP_JuMP_model(Q,R,A,B, nt; lvar = lvar, uvar= uvar, Qf = Qf)
+qpulbf = get_QM(Q, R, A, B, nt; lvar = lvar, uvar=uvar, Qf = Qf)
+
+# Solve new problem with Qf matrix
+optimize!(mulbf)
+ipopt_ulbf  = ipopt(qpulbf)
+madnlp_ulbf = madnlp(qpulbf)
+
+# Test results
+@test abs(objective_value(mulbf) - ipopt_ulbf.objective)  < 1e-7
+@test abs(objective_value(mulbf) - madnlp_ulbf.objective) < 1e-7
+
+
+# Test new problem with x0
+x0 = lvar[1:2] .+ .5
+
+mulbx0  = build_QP_JuMP_model(Q,R,A,B, nt; lvar = lvar, uvar= uvar, x0 = x0)
+qpulbx0 = get_QM(Q, R, A, B, nt; lvar = lvar, uvar=uvar, x0=x0)
+
+# Solve new problem with x0
+optimize!(mulbx0)
+ipopt_ulbx0   = ipopt(qpulbx0)
+madnlp_ulbx0  = madnlp(qpulbx0)
+
+# Test results
+@test abs(objective_value(mulbx0) - ipopt_ulbx0.objective)  < 1e-7
+@test abs(objective_value(mulbx0) - madnlp_ulbx0.objective) < 1e-7
+@test abs(madnlp_ulbx0.objective  - ipopt_ulbx0.objective) < 1e-7
+
+
+
```

Comment:
Let's remove 104-end. In principle, tests shouldn't fail as long as we have a correct implementation, though we want it to cover as many edge cases as possible. What we have here is not an edge case, since we may actually have an infeasible problem, due to the randomness. If we want to test that our package correctly identifies the infeasibility, we need to *intentionally* put an infeasible problem and see if the solver actually exits with infeasibility.